### PR TITLE
Perf: Avoid lambda expression compilation for Singleton scoped services

### PR DESCRIPTION
**Pull request by mnltejaswini**
The complete details for this PR are included in the #425.

### CODE REVIEW

**Comment by mnltejaswini** [comment]
@rynowak

File: src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs [hunk 1]
```diff
@@ -73,7 +73,12 @@ public object GetService(Type serviceType)
             var callCount = 0;
             return provider =>
             {
-                if (Interlocked.Increment(ref callCount) == 2)
+                // PERF: Singleton services once created will always be available in the lookup dictionary.
+                // Hence it is not needed to create an lambda expression for looking up in the dictionary.
+                // This change reduces the memory allocations during the start up because 
+                // lambda expression compilation involves expensive code access security checks for Desktop CLR.
+                var singletonCallSite = callSite as SingletonCallSite;
```
Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Would there be any benefit in not creating the `Func` for singletons? i.e. have `RealizedServices` store the instance when it's a singleton and `Func` otherwise?

File: src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs [hunk 2]
```diff
@@ -362,7 +368,15 @@ public override object Invoke(ServiceProvider provider)
...
-                return base.Build(Expression.Field(provider, "_root"));
+                // PERF: Singleton services once created will always be available in the lookup dictionary
+                // Hence it is cheaper to directly call its Invoke method that will do the lookup rather than 
+                // building an expression tree that does the same.
+                // This change reduces the memory allocations during the start up because, expressions accessing private members
+                // involves expensive code access security checks for Desktop CLR.
+                return Expression.Call(
+                    Expression.Constant(this, typeof(SingletonCallSite)),
+                    InvokeMethodInfo,
+                    provider);
```
Comment (rynowak) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: This still appears to involve generating an expression - why is that needed?

File: src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: That would double the size of the service table.

File: src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs (hunk 2 above)
Comment (mnltejaswini) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Yes. This change is for other call site types that reference singletons

Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Also make a note that the singleton would have already been registered for disposal at this point. (We're removing the call to `base.Build`).

File: src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Didn't see a particularly cleaner pattern removing the polymorphism. It's used as a way to reuse code here and changing it would just involve a lot of code re-organization without any specific benefits. That said, it would make sense to move this check to `CreateServiceAccessor` and special case top level singletons outside the closure.

Comment (lodejard) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: The only changes would be for `IServiceProvider.GetService` directly for a singleton. Plus it happens out-of-band (in a Task, not in the main-line code path). Is this a realistic optimization?

Comment (rynowak) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Expression compilation allocates to an extreme degree on desktop due to CAS. https://github.com/aspnet/DependencyInjection/issues/425 This is causing a lot of heap-size growth at startup. As far as the breakdown between the impact of this change vs the change to `SingletonCallSite.Build`, @mnltejaswini will have to answer

Comment (lodejard) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Okay... Then how about moving the down-cast into the Task.Run body? if callSite is singleton, `table.RealizedService[serviceType] = callSite.Invoke;` else compile expression?

File: src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs (hunk 2 above)
Comment (lodejard) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Is the memory allocation proportional to the number of expression nodes, or to the number of calls to lambda compile?

Assuming it's proportional to the number of calls to compile - I don't think you need to change this method at all if the top-level Build isn't called for direct calls to GetService of singletons.

Comment (mnltejaswini) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Although avoiding top level Build optimizes the direct calls to GetService of singletons, this change to Build optimizes the GetService of other call sites that have singletons in their expression graph. I will get some numbers for each of the change separately for the customer app.

Comment (mnltejaswini) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: Here are the numbers measured for startup with each change separately. The maximum benefit comes from changing the Build method of SIngletonCallSite

| Metric   |      Total allocations       |  Total private bytes of managed Heap |
|----------|:-------------:|------:|
| Avoid Build for direct calls to get Singleton services |  239MB | 145MB |
| Changing Build for Singleton call site to reduce the nodes in the expression tree for dependent services|    115MB   |   95 MB |

Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: To @pranavkm's point about singletons being added to the list of disposables, how does this work in those cases? Is it because invoke gets called before this ever does?

Comment (halter73) [inline src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs]: This actually looks pretty safe and reasonable. Calling `Invoke()` instead of doing the equivalent action directly in the expression tree isn't really that costly since we'll always be pulling the instance out of a dictionary anyway instead of doing any complicated reflection, constructor selection, etc... If calling `Invoke()` avoids costly CAS-related allocations, that seems better to me than some other sugested alternatives such as making `_root` public.

@davidfowl The new compiled expression returned by this method, `SingletonCallSite.Build`, calls into `ScopedCallSite.Invoke` which would add a newly created singleton to `_resolvedServices` (used in part to track disposables) if it wasn't there already. Even though it's not really possible as @pranavkm mentions for the singleton not to be in the dictionary yet, the code would still be correct (albeit slow) if that wasn't the case.

**Comment by halter73** [comment]
:shipit: This doesn't preclude doing me extensive changes (e.g. memoization) later, and it certainly seems way better than what we have now on desktop.

**Comment by pakrym** [comment]
This PR fixes allocations for singletons but not scoped services, avoiding CAS checks when generating expressions would fix both.

**Comment by rynowak** [comment]
Per discussion, @pakrym is going to follow up on this, and possibly produce a more all-encompassing fix.

**Comment by pakrym** [comment]
Closing. Issue this PR addressed was fixed in https://github.com/aspnet/DependencyInjection/pull/429
